workflow/airflow_libero/dags/ingest_article_xml_dag.py

```diff
@@ -0,0 +1,125 @@
+from datetime import timedelta
+import os
+import json
+import sys
+
+import airflow
+from airflow import DAG
+from airflow.operators.python_operator import PythonOperator
+from bs4 import BeautifulSoup
+import requests
+
+# HACK for DAG example usage without creating a load of package structures
+# would break this out, outside of example usage
+sys.path.append('..')
+from airflow_libero.operators import EventEmittingPythonOperator
+
+DIR_PATH = 'data'
+
+
+def _download_article_xml(*args, **kwargs):
+    """Download some article xml and store in a public location
+
+    :return: str
+    """
+
+    input_data = kwargs['dag_run'].conf.get('input_data')
+
+    if input_data:
+        data_dir = input_data.get('data_dir')
+        file_urls = input_data.get('urls')
+
+        if not os.path.isdir(data_dir):
+            os.mkdir(data_dir)
+
+        if file_urls:
+            for url in file_urls:
+                response = requests.get(url=url)
+
+                file_name = url.split('/')[-1]
+
+                with open(os.path.join(data_dir, file_name), 'w') as xml_file:
+                    xml_file.write(response.text)
+
+
+def _extract_asset_uris(*args, **kwargs):
+    """Download some article xml and extract asset URIs from XML.
+
+    :return: str
+    """
+
+    input_data = kwargs['dag_run'].conf.get('input_data')
+
+    if input_data:
+        file_urls = input_data.get('urls')
+
+        uris = []
+
+        if file_urls:
+            for url in file_urls:
+                response = requests.get(url=url)
+                xml = BeautifulSoup(response.text, 'lxml-xml')
+                uris += [var.contents[0] for var in xml.find_all('source')
+                         if var.attrs['media-type'] == 'image/tiff']
+
+        print(uris)
+
+        kwargs['ti'].xcom_push('uris', json.dumps({'uris': uris}))
+
+
+def _download_assets(*args, **kwargs):
+    """Download some static assets.
+
+    :return: str
+    """
+
+    input_data = kwargs['dag_run'].conf.get('input_data')
+    uris = kwargs['ti'].xcom_pull(task_ids=None, key='uris')
+
+    if input_data and uris:
+        data_dir = input_data.get('data_dir')
+
+        uris = json.loads(uris.replace("'", '"'))
+
+        for uri in uris['uris']:
+            response = requests.get(url=uri)
+
+            file_name = uri.split('/')[-1]
+
+            with open(os.path.join(data_dir, file_name), 'wb') as asset_file:
+                asset_file.write(response.content)
+
+
+default_args = {
+    'owner': 'airflow',
+    'depends_on_past': False,
+    'start_date': airflow.utils.dates.days_ago(2),
```

Comment:
hackish way to get it to run immediately I guess - shows the bias towards scheduling